File: extension/backup.js
'use strict';

const sanitizeFileName = (name) => {
  const cleaned = (name || '').replace(/[\\/:*?"<>|\n\r]/g, '_').trim();
  return cleaned || 'untitled';
};

const localDateKey = (timestamp) => {
  const date = new Date(timestamp);
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
};

const mimeExt = (mime) => {
  const table = {
    'image/png': 'png',
    'image/jpeg': 'jpg',
    'image/gif': 'gif',
    'image/webp': 'webp',
    'video/mp4': 'mp4',
    'video/webm': 'webm',
  };
  return table[mime] || 'bin';
};

const COMPRESSED_MIME = /^(image|video|audio)\//;

const zipAsync = (files) =>
  new Promise((resolve, reject) => {
    fflate.zip(files, { level: 6 }, (err, data) => (err ? reject(err) : resolve(data)));
  });

const unzipAsync = (bytes) =>
  new Promise((resolve, reject) => {
    fflate.unzip(bytes, (err, files) => (err ? reject(err) : resolve(files)));
  });

const exportZip = async () => {
  const folders = await getActiveFolders();
  const allNotes = await db.notes.toArray();
  const activeNotes = allNotes.filter((note) => note.deletedAt == null);
  const blobs = await db.blobs.toArray();
  const blobById = new Map(blobs.map((record) => [record.id, record]));

  const prefixCount = new Map();
  for (const record of blobs) {
    const base = record.sha256.slice(0, 8);
    prefixCount.set(base, (prefixCount.get(base) || 0) + 1);
  }
  const usedNames = new Set();
  const assetNames = new Map();
  for (const record of blobs) {
    let base = record.sha256.slice(0, 8);
    if (prefixCount.get(base) > 1) {
      base = record.sha256.slice(0, 16);
    }
    const ext = mimeExt(record.mime);
    let name = `${base}.${ext}`;
    let counter = 2;
    while (usedNames.has(name)) {
      name = `${base}-${counter}.${ext}`;
      counter++;
    }
    usedNames.add(name);
    assetNames.set(record.id, name);
  }

  const files = {};

  for (const record of blobs) {
    if (!record.blob) {
      continue;
    }
    const buffer = await record.blob.arrayBuffer();
    const bytes = new Uint8Array(buffer);
    const key = `assets/${assetNames.get(record.id)}`;
    files[key] = COMPRESSED_MIME.test(record.mime || '') ? [bytes, { level: 0 }] : bytes;
  }

  const notesByFolder = new Map();
  for (const note of activeNotes) {
    if (!notesByFolder.has(note.folderId)) {
      notesByFolder.set(note.folderId, []);
    }
    notesByFolder.get(note.folderId).push(note);
  }

  for (const folder of folders) {
    const folderNotes = notesByFolder.get(folder.id) || [];
    if (folderNotes.length === 0) {
      continue;
    }
    const dir = sanitizeFileName(folder.name);

    const byDay = new Map();
    for (const note of folderNotes) {
      const day = localDateKey(note.createdAt);
      if (!byDay.has(day)) {
        byDay.set(day, []);
      }
      byDay.get(day).push(note);
    }

    for (const [day, dayNotes] of byDay) {
      dayNotes.sort((a, b) => (a.createdAt || 0) - (b.createdAt || 0));
      let markdown = '';
      for (const note of dayNotes) {
        let content = note.content || '';
        content = content.replace(/blob:(\d+)/g, (match, idText) => {
          const record = blobById.get(Number(idText));
          if (!record) {
            return match;
          }
          return `../assets/${assetNames.get(record.id)}`;
        });
        if (note.sourceUrl) {
          const label = '출처';
          const source = `${note.sourceTitle || ''} ${note.sourceUrl}`.trim();
          content += `\n\n> ${label}: ${source}\n`;
        }
        markdown += `${content.trim()}\n\n---\n\n`;
      }
      files[`${dir}/${day}.md`] = textEncoder.encode(markdown);
    }
  }

  const backup = {
    app: 'memo-stream',
    version: 1,
    exportedAt: new Date().toISOString(),
    folders: folders.map((folder) => ({
      id: folder.id,
      name: folder.name,
      order: folder.order,
      pinned: folder.pinned,
      createdAt: folder.createdAt,
    })),
    notes: activeNotes.map((note) => ({
      id: note.id,
      folderId: note.folderId,
      content: note.content,
      blobIds: note.blobIds || [],
      sourceUrl: note.sourceUrl,
      sourceTitle: note.sourceTitle,
      pinned: note.pinned,
      createdAt: note.createdAt,
      editedAt: note.editedAt,
    })),
    blobs: blobs.map((record) => ({
      id: record.id,
      sha256: record.sha256,
      mime: record.mime,
      name: record.name || '',
      size: record.size,
      thumbSize: record.thumbSize,
      width: record.width,
      height: record.height,
      refCount: record.refCount,
      createdAt: record.createdAt,
    })),
  };
  files['backup.json'] = textEncoder.encode(JSON.stringify(backup));

  const data = await zipAsync(files);
  return { data, name: `memo-export-${localDateKey(Date.now())}.zip` };
};

const unzipData = (data) => {
  let bytes = data;
  if (!(data instanceof Uint8Array)) {
    bytes = new Uint8Array(data);
  }
  return unzipAsync(bytes);
};

const findAssetKey = (files, sha256) => {
  const short = (sha256 || '').slice(0, 8);
  const long = (sha256 || '').slice(0, 16);
  for (const key of Object.keys(files)) {
    if (!key.startsWith('assets/')) {
      continue;
    }
    const base = key.slice('assets/'.length).split('.')[0];
    if (base === short || base === long || base.startsWith(`${short}-`)) {
      return key;
    }
  }
  return null;
};

const importFromBackup = async (backup, files) => {
  let blobsImported = 0;
  let blobsReused = 0;
  const blobIdMap = new Map();

  for (const entry of backup.blobs || []) {
    const assetKey = findAssetKey(files, entry.sha256);
    if (!assetKey) {
      continue;
    }
    const mime = entry.mime || 'application/octet-stream';
    const blob = new Blob([files[assetKey]], { type: mime });
    const result = await findOrInsertBlob({
      blob,
      mime,
      thumb: null,
      name: entry.name || '',
      size: entry.size || blob.size,
      thumbSize: 0,
      width: entry.width || 0,
      height: entry.height || 0,
    });
    blobIdMap.set(entry.id, result.id);
    if (result.reused) {
      blobsReused++;
    } else {
      blobsImported++;
    }
  }

  const existingFolders = (await db.folders.toArray()).filter((row) => row.deletedAt == null);
  const folderIdMap = new Map();
  let foldersCreated = 0;
  for (const folder of backup.folders || []) {
    const found = existingFolders.find((row) => row.name === folder.name);
    if (found) {
      folderIdMap.set(folder.id, found.id);
      continue;
    }
    const createdAt = folder.createdAt || Date.now();
    const id = await db.folders.add({
      uid: newUid(),
      name: folder.name,
      order: folder.order || 0,
      pinned: !!folder.pinned,
      createdAt,
      deletedAt: null,
      updatedAt: Date.now(),
    });
    existingFolders.push({ id, name: folder.name, deletedAt: null });
    folderIdMap.set(folder.id, id);
    foldersCreated++;
  }

  let notesCreated = 0;
  for (const note of backup.notes || []) {
    let folderId = folderIdMap.get(note.folderId);
    if (folderId == null) {
      folderId = await ensureInbox();
    }
    const blobIds = (note.blobIds || [])
      .map((oldId) => blobIdMap.get(oldId))
      .filter((id) => id != null);
    const content = (note.content || '').replace(/blob:(\d+)/g, (match, idText) => {
      const newId = blobIdMap.get(Number(idText));
      if (newId == null) {
        return match;
      }
      return `blob:${newId}`;
    });
    await db.notes.add({
      uid: newUid(),
      folderId,
      content,
      blobIds,
      sourceUrl: note.sourceUrl || null,
      sourceTitle: note.sourceTitle || null,
      pinned: !!note.pinned,
      createdAt: note.createdAt || Date.now(),
      editedAt: note.editedAt || null,
      deletedAt: null,
      updatedAt: Date.now(),
    });
    for (const id of blobIds) {
      await incrementRefCount(id);
    }
    notesCreated++;
  }

  for (const id of blobIdMap.values()) {
    await decrementRefCount(id);
  }

  return {
    folders: foldersCreated,
    notes: notesCreated,
    blobs: blobsImported,
    reusedBlobs: blobsReused,
  };
};

const importFromMarkdown = async (files) => {
  const keys = Object.keys(files);
  const isImageFile = (key) => /\.(png|jpe?g|gif|webp)$/i.test(key);
  const mimeByExt = {
    png: 'image/png',
    jpg: 'image/jpeg',
    jpeg: 'image/jpeg',
    gif: 'image/gif',
    webp: 'image/webp',
  };

  const imageMap = new Map();
  let blobsImported = 0;
  let blobsReused = 0;
  for (const key of keys) {
    if (!isImageFile(key)) {
      continue;
    }
    const fileName = key.split('/').pop().toLowerCase();
    if (imageMap.has(fileName)) {
      continue;
    }
    const ext = fileName.split('.').pop();
    const mime = mimeByExt[ext] || 'application/octet-stream';
    const blob = new Blob([files[key]], { type: mime });
    const result = await findOrInsertBlob({
      blob,
      mime,
      thumb: null,
      name: fileName,
      size: blob.size,
      thumbSize: 0,
      width: 0,
      height: 0,
    });
    imageMap.set(fileName, result.id);
    if (result.reused) {
      blobsReused++;
    } else {
      blobsImported++;
    }
  }

  let foldersCreated = 0;
  let notesCreated = 0;
  const folderCache = new Map();

  const getFolderForDir = async (dir) => {
    if (!dir || dir === 'assets') {
      return ensureInbox();
    }
    if (folderCache.has(dir)) {
      return folderCache.get(dir);
    }
    const existing = (await db.folders.toArray()).filter((folder) => folder.deletedAt == null);
    const found = existing.find((folder) => folder.name === dir);
    let folderId = null;
    if (found) {
      folderId = found.id;
    } else {
      folderId = await createFolder(dir);
      foldersCreated++;
    }
    folderCache.set(dir, folderId);
    return folderId;
  };

  for (const key of keys) {
    if (!/\.md$/i.test(key)) {
      continue;
    }
    if (key.startsWith('assets/')) {
      continue;
    }
    const parts = key.split('/');
    let dir = '';
    if (parts.length > 1) {
      dir = parts[0];
    }
    const folderId = await getFolderForDir(dir);

    const raw = textDecoder.decode(files[key]);
    const usedIds = new Set();
    const content = raw.replace(/!\[[^\]]*\]\(([^)]+)\)/g, (match, ref) => {
      const cleanRef = ref.trim().replace(/^\.?\//, '');
      const fileName = cleanRef.split('/').pop().toLowerCase();
      const id = imageMap.get(fileName);
      if (id == null) {
        return match;
      }
      usedIds.add(id);
      return `![](blob:${id})`;
    });

    const now = Date.now();
    await db.notes.add({
      uid: newUid(),
      folderId,
      content,
      blobIds: [...usedIds],
      sourceUrl: null,
      sourceTitle: null,
      pinned: false,
      createdAt: now,
      editedAt: null,
      deletedAt: null,
      updatedAt: now,
    });
    for (const id of usedIds) {
      await incrementRefCount(id);
    }
    notesCreated++;
  }

  for (const id of imageMap.values()) {
    await decrementRefCount(id);
  }

  return {
    folders: foldersCreated,
    notes: notesCreated,
    blobs: blobsImported,
    reusedBlobs: blobsReused,
  };
};

const importFromZip = async (data) => {
  const files = await unzipData(data);
  if (files['backup.json']) {
    const backup = JSON.parse(textDecoder.decode(files['backup.json']));
    return importFromBackup(backup, files);
  }
  return importFromMarkdown(files);
};

File: app/app/src/main/java/com/example/memostream/data/Backup.kt
package com.example.memostream.data

import java.io.File
import java.io.InputStream
import java.io.OutputStream
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream
import org.json.JSONArray
import org.json.JSONObject

private val INVALID_NAME = Regex("""[\\/:*?"<>|\n\r]""")

fun sanitizeFileName(name: String): String =
    INVALID_NAME.replace(name, "_").trim().ifEmpty {
        "untitled"
    }

fun localDateKey(timestamp: Long): String =
    SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date(timestamp))

private fun extOf(mime: String): String = when (mime) {
    "image/png" -> "png"
    "image/jpeg" -> "jpg"
    "image/gif" -> "gif"
    "image/webp" -> "webp"
    "video/mp4" -> "mp4"
    "video/webm" -> "webm"
    else -> "bin"
}

data class ImportResult(val folders: Int, val notes: Int, val blobs: Int, val reusedBlobs: Int)

class Backup(private val repo: MemoRepository) {
    fun exportName(): String = "memo-export-${localDateKey(System.currentTimeMillis())}.zip"

    suspend fun exportTo(output: OutputStream) {
        val folders = repo.folders(true)
        val notes = repo.allNotes().filter {
            it.deletedAt == null
        }
        val blobs = repo.blobs()

        val prefixCount = HashMap<String, Int>()
        blobs.forEach { record ->
            val base = record.sha256.take(8)
            prefixCount[base] = (prefixCount[base] ?: 0) + 1
        }
        val used = HashSet<String>()
        val assetNames = HashMap<Long, String>()
        blobs.forEach { record ->
            var base = record.sha256.take(8)
            if ((prefixCount[base] ?: 0) > 1) {
                base = record.sha256.take(16)
            }
            val ext = extOf(record.mime)
            var name = "$base.$ext"
            var counter = 2
            while (used.contains(name)) {
                name = "$base-$counter.$ext"
                counter++
            }
            used.add(name)
            assetNames[record.id] = name
        }

        ZipOutputStream(output).use { zip ->
            blobs.forEach { record ->
                val file = repo.files.blob(record.sha256)
                if (!file.exists()) {
                    return@forEach
                }
                zip.putNextEntry(ZipEntry("assets/${assetNames[record.id]}"))
                file.inputStream().use {
                    it.copyTo(zip)
                }
                zip.closeEntry()
            }

            val byFolder = notes.groupBy {
                it.folderId
            }
            folders.forEach { folder ->
                val folderNotes = byFolder[folder.id].orEmpty()
                if (folderNotes.isEmpty()) {
                    return@forEach
                }
                val dir = sanitizeFileName(folder.name)
                folderNotes.groupBy {
                    localDateKey(it.createdAt)
                }.forEach { (day, dayNotes) ->
                    val markdown = StringBuilder()
                    dayNotes.sortedBy {
                        it.createdAt
                    }.forEach { note ->
                        var content = note.content
                        Regex("""blob:(\d+)""").findAll(note.content).forEach { match ->
                            val id = match.groupValues[1].toLongOrNull() ?: return@forEach
                            val asset = assetNames[id] ?: return@forEach
                            content = content.replace("blob:$id", "../assets/$asset")
                        }
                        if (note.sourceUrl != null) {
                            val source = listOfNotNull(note.sourceTitle, note.sourceUrl)
                                .joinToString(" ").trim()
                            content += "\n\n> 출처: $source\n"
                        }
                        markdown.append(content.trim()).append("\n\n---\n\n")
                    }
                    zip.putNextEntry(ZipEntry("$dir/$day.md"))
                    zip.write(markdown.toString().toByteArray())
                    zip.closeEntry()
                }
            }

            val payload = JSONObject().apply {
                put("app", "memo-stream")
                put("version", 2)
                put("exportedAt", System.currentTimeMillis())
                put("folders", JSONArray().apply {
                    folders.forEach { folder ->
                        put(JSONObject().apply {
                            put("id", folder.id)
                            put("uid", folder.uid)
                            put("name", folder.name)
                            put("order", folder.order)
                            put("pinned", folder.pinned)
                            put("createdAt", folder.createdAt)
                        })
                    }
                })
                put("notes", JSONArray().apply {
                    notes.forEach { note ->
                        put(JSONObject().apply {
                            put("id", note.id)
                            put("folderId", note.folderId ?: JSONObject.NULL)
                            put("content", note.content)
                            put("blobIds", JSONArray(note.blobIds))
                            put("sourceUrl", note.sourceUrl ?: JSONObject.NULL)
                            put("sourceTitle", note.sourceTitle ?: JSONObject.NULL)
                            put("pinned", note.pinned)
                            put("createdAt", note.createdAt)
                            put("editedAt", note.editedAt ?: JSONObject.NULL)
                        })
                    }
                })
                put("blobs", JSONArray().apply {
                    blobs.forEach { record ->
                        put(JSONObject().apply {
                            put("id", record.id)
                            put("sha256", record.sha256)
                            put("mime", record.mime)
                            put("name", record.name)
                            put("size", record.size)
                            put("width", record.width)
                            put("height", record.height)
                            put("loop", record.loop)
                        })
                    }
                })
            }
            zip.putNextEntry(ZipEntry("backup.json"))
            zip.write(payload.toString().toByteArray())
            zip.closeEntry()
        }
    }

    suspend fun importFrom(input: InputStream): ImportResult {
        val entries = HashMap<String, ByteArray>()
        ZipInputStream(input).use { zip ->
            while (true) {
                val entry = zip.nextEntry ?: break
                if (!entry.isDirectory) {
                    entries[entry.name] = zip.readBytes()
                }
                zip.closeEntry()
            }
        }
        val manifest = entries["backup.json"]
        return if (manifest != null) {
            importBackup(JSONObject(String(manifest)), entries)
        } else {
            importMarkdown(entries)
        }
    }

    private fun findAsset(entries: Map<String, ByteArray>, sha256: String): ByteArray? {
        val short = sha256.take(8)
        val long = sha256.take(16)
        return entries.entries.firstOrNull { (key, _) ->
            if (!key.startsWith("assets/")) {
                return@firstOrNull false
            }
            val base = key.removePrefix("assets/").substringBefore('.')
            base == short || base == long || base.startsWith("$short-")
        }?.value
    }

    private suspend fun insertBytes(bytes: ByteArray, mime: String, name: String, loop: Boolean): Pair<Long, Boolean> {
        val sha = java.security.MessageDigest.getInstance("SHA-256")
            .digest(bytes).joinToString("") {
                "%02x".format(it)
            }
        repo.blobBySha(sha)?.let {
            repo.incrementRef(it.id)
            return it.id to true
        }
        repo.files.blob(sha).writeBytes(bytes)
        val id = repo.insertBlob(
            BlobRecord(
                sha256 = sha,
                mime = mime,
                name = name,
                size = bytes.size.toLong(),
                loop = loop,
                refCount = 1,
                createdAt = System.currentTimeMillis(),
            )
        )
        return id to false
    }

    private suspend fun importBackup(payload: JSONObject, entries: Map<String, ByteArray>): ImportResult {
        var imported = 0
        var reused = 0
        val blobIdMap = HashMap<Long, Long>()

        val blobs = payload.optJSONArray("blobs") ?: JSONArray()
        for (i in 0 until blobs.length()) {
            val row = blobs.getJSONObject(i)
            val bytes = findAsset(entries, row.optString("sha256")) ?: continue
            val (id, wasReused) = insertBytes(
                bytes,
                row.optString("mime").ifEmpty {
                    "application/octet-stream"
                },
                row.optString("name"),
                row.optBoolean("loop"),
            )
            blobIdMap[row.optLong("id")] = id
            if (wasReused) {
                reused++
            } else {
                imported++
            }
        }

        val existing = repo.folders(true).toMutableList()
        val folderIdMap = HashMap<Long, Long>()
        var foldersCreated = 0
        val folders = payload.optJSONArray("folders") ?: JSONArray()
        for (i in 0 until folders.length()) {
            val row = folders.getJSONObject(i)
            val name = row.optString("name")
            val found = existing.firstOrNull {
                it.name == name
            }
            if (found != null) {
                folderIdMap[row.optLong("id")] = found.id
                continue
            }
            val id = repo.createFolder(name)
            repo.folder(id)?.let {
                existing.add(it)
            }
            folderIdMap[row.optLong("id")] = id
            foldersCreated++
        }

        var notesCreated = 0
        val notes = payload.optJSONArray("notes") ?: JSONArray()
        for (i in 0 until notes.length()) {
            val row = notes.getJSONObject(i)
            val folderId = folderIdMap[row.optLong("folderId")] ?: repo.ensureInbox()
            val oldIds = row.optJSONArray("blobIds") ?: JSONArray()
            val newIds = (0 until oldIds.length()).mapNotNull {
                blobIdMap[oldIds.optLong(it)]
            }
            var content = row.optString("content")
            blobIdMap.forEach {
                (old, new) -> content = content.replace("blob:$old)", "blob:$new)")
            }
            repo.insertNote(
                Note(
                    uid = newUid(),
                    folderId = folderId,
                    content = content,
                    blobIds = newIds,
                    sourceUrl = row.optString("sourceUrl").takeIf {
                        !row.isNull("sourceUrl")
                    },
                    sourceTitle = row.optString("sourceTitle").takeIf {
                        !row.isNull("sourceTitle")
                    },
                    pinned = row.optBoolean("pinned"),
                    createdAt = row.optLong("createdAt").takeIf {
                        it > 0
                    } ?: System.currentTimeMillis(),
                    editedAt = row.optLong("editedAt").takeIf {
                        !row.isNull("editedAt") && it > 0
                    },
                    updatedAt = System.currentTimeMillis(),
                )
            )
            newIds.forEach {
                repo.incrementRef(it)
            }
            notesCreated++
        }

        blobIdMap.values.forEach {
            repo.decrementRef(it)
        }
        return ImportResult(foldersCreated, notesCreated, imported, reused)
    }

    private suspend fun importMarkdown(entries: Map<String, ByteArray>): ImportResult {
        val mimeByExt = mapOf(
            "png" to "image/png", "jpg" to "image/jpeg", "jpeg" to "image/jpeg",
            "gif" to "image/gif", "webp" to "image/webp",
        )
        val imageMap = HashMap<String, Long>()
        var imported = 0
        var reused = 0
        entries.forEach { (key, bytes) ->
            val fileName = key.substringAfterLast('/').lowercase()
            val ext = fileName.substringAfterLast('.', "")
            val mime = mimeByExt[ext] ?: return@forEach
            if (imageMap.containsKey(fileName)) {
                return@forEach
            }
            val loop = mime == "image/gif" && gifIsAnimated(bytes)
            val (id, wasReused) = insertBytes(bytes, mime, fileName, loop)
            imageMap[fileName] = id
            if (wasReused) {
                reused++
            } else {
                imported++
            }
        }

        var foldersCreated = 0
        var notesCreated = 0
        val folderCache = HashMap<String, Long>()

        suspend fun folderFor(dir: String): Long {
            if (dir.isEmpty() || dir == "assets") {
                return repo.ensureInbox()
            }
            folderCache[dir]?.let {
                return it
            }
            val found = repo.folders(true).firstOrNull {
                it.name == dir
            }
            val id = if (found != null) found.id else {
                foldersCreated++
                repo.createFolder(dir)
            }
            folderCache[dir] = id
            return id
        }

        entries.forEach { (key, bytes) ->
            if (!key.endsWith(".md", ignoreCase = true) || key.startsWith("assets/")) {
                return@forEach
            }
            val parts = key.split('/')
            val folderId = folderFor(if (parts.size > 1) parts[0] else "")
            val usedIds = LinkedHashSet<Long>()
            val content = Regex("""!\[[^\]]*]\(([^)]+)\)""").replace(String(bytes)) { match ->
                val fileName = match.groupValues[1].trim().substringAfterLast('/').lowercase()
                val id = imageMap[fileName] ?: return@replace match.value
                usedIds.add(id)
                "![](blob:$id)"
            }
            val now = System.currentTimeMillis()
            repo.insertNote(
                Note(
                    uid = newUid(),
                    folderId = folderId,
                    content = content,
                    blobIds = usedIds.toList(),
                    createdAt = now,
                    updatedAt = now,
                )
            )
            usedIds.forEach {
                repo.incrementRef(it)
            }
            notesCreated++
        }

        imageMap.values.forEach {
            repo.decrementRef(it)
        }
        return ImportResult(foldersCreated, notesCreated, imported, reused)
    }
}
